source/org/zfin/expression/repository/HibernateExpressionRepository.java

```diff
@@ -1390,34 +1380,43 @@ public boolean pileStructureExists(ExpressedTermDTO expressedTerm, String public
         }
 
         Session session = HibernateUtil.currentSession();
+        String hql = """
+            from ExpressionStructure
+            """;
 
-        Criteria crit = session.createCriteria(ExpressionStructure.class);
+        List<String> hqlClauses = new ArrayList<>();
+        HashMap<String, Object> parameterMap = new HashMap<>();
+
+        hqlClauses.add("superterm.termName = :superterm");
+        parameterMap.put("superterm", supertermName);
+        hqlClauses.add("expressionFound = :expressionFound");
+        parameterMap.put("expressionFound", expressedTerm.isExpressionFound());
+        hqlClauses.add("publication.zdbID = :publicationZdbID");
+        parameterMap.put("publicationZdbID", publicationID);
         if (expressedTerm.getEntity().getSubTerm() != null) {
             OntologyDTO subtermOntology = expressedTerm.getEntity().getSubTerm().getOntology();
             if (subtermOntology == null) {
                 throw new NullPointerException("No subterm ontology provided.");
             }
-            Criteria subterm = crit.createCriteria("subterm");
-            subterm.add(Restrictions.eq("termName", expressedTerm.getEntity().getSubTerm().getTermName()));
+            hqlClauses.add("subterm = :subterm");
+            parameterMap.put("subterm", expressedTerm.getEntity().getSubTerm());
         } else {
-            crit.add(Restrictions.isNull("subterm"));
+            hqlClauses.add("subterm is null");
         }
         if (expressedTerm.getQualityTerm() != null) {
             if (expressedTerm.getQualityTerm().getTerm() != null) {
-                crit.add(Restrictions.eq("eapQualityTerm.zdbID", expressedTerm.getQualityTerm().getTerm().getZdbID()));
+                hqlClauses.add("eapQualityTerm.zdbID = :zdbID");
+                parameterMap.put("zdbID", expressedTerm.getQualityTerm().getTerm().getZdbID());
             }
-            crit.add(Restrictions.eq("tag", expressedTerm.getQualityTerm().getTag()));
+            hqlClauses.add("tag = :tag");
+            parameterMap.put("tag", expressedTerm.getQualityTerm().getTag());
         } else {
-            crit.add(Restrictions.isNull(("eapQualityTerm")));
+            hqlClauses.add("eapQualityTerm is null");
         }
-        crit.add(Restrictions.eq("expressionFound", expressedTerm.isExpressionFound()));
-        Criteria publication = crit.createCriteria("publication");
-        publication.add(Restrictions.eq("zdbID", publicationID));
-        Criteria superterm = crit.createCriteria("superterm");
-        superterm.add(Restrictions.eq("termName", supertermName));
-
-        List list = crit.list();
-        return list != null && !list.isEmpty();
+        hql += " where " + String.join(" and ", hqlClauses);
+        Query<ExpressionStructure> query = session.createQuery(hql, ExpressionStructure.class);
+        parameterMap.forEach(query::setParameter);
+        return query.list() != null && !query.list().isEmpty();

```

Comment:
Does this run the query twice? (once for `query.list() != null` and once for `!query.list().isEmpty()`?)